client/allocrunner/taskrunner/logmon_hook.go

```diff
@@ -95,7 +97,37 @@ func reattachConfigFromHookData(data map[string]string) (*plugin.ReattachConfig,
 func (h *logmonHook) Prestart(ctx context.Context,
 	req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
 
-	// Attempt to reattach to logmon
+	tries := 0
+	for {
+		err := h.prestartOneLoop(ctx, req)
+		if err == bstructs.ErrPluginShutdown {
+			h.logger.Warn("logmon shutdown while making request", "error", err)
+
+			if tries > 3 {
+				return err
+			}
+
+			// retry after killing process and ensure we start a new logmon process
+			tries++
+			h.logmonPluginClient.Kill()
+			time.Sleep(1 * time.Second)
+			continue
+		} else if err != nil {
+			return err
+		}
+
+		rCfg := pstructs.ReattachConfigFromGoPlugin(h.logmonPluginClient.ReattachConfig())
+		jsonCfg, err := json.Marshal(rCfg)
+		if err != nil {
+			return err
+		}
+		resp.State = map[string]string{logmonReattachKey: string(jsonCfg)}
+		return nil
+	}
+}
+
+func (h *logmonHook) prestartOneLoop(ctx context.Context, req *interfaces.TaskPrestartRequest) error {
+	// attach to a running logmon if state indicates one
```

Comment:
I opted not to change the logic here as the logic is somewhat brittle and I don't want to make it worse.  But the idea is that if grpc call fails with shutting down, `h.logmonPluginClient` would be non-nill and would be marked as `Exited`, so we'll create a new logmon instance on retries.